src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Package/Update/PackageUpdateCommandRunner.cs

```diff
@@ -0,0 +1,261 @@
+
+            await Task.Delay(1, cancellationToken);
+
+            // 1. Get DGSpec for project/solution
+            string settingsRoot = Directory.Exists(args.Project) ? args.Project : Path.GetDirectoryName(args.Project)!;
+            ISettings settings = Settings.LoadDefaultSettings(settingsRoot);
+
+            var dgSpec = dGSpecFactory.GetDependencyGraphSpec(args.Project);
+
+            if (dgSpec.Restore.Count > 1)
+            {
+                logger.LogMinimal("Unsupported: Updating more than one project is not yet supported", ConsoleColor.Red);
+                return ExitCodes.Error;
+            }
+
+            // 2. Find suitable version of package(s) to update
+            // Source provider will be needed to find the package version and to restore, so create it here.
+            CachingSourceProvider sourceProvider = new CachingSourceProvider(new PackageSourceProvider(settings));
+            using SourceCacheContext sourceCacheContext = new();
+            (PackageDependency? packageToUpdate, List<NuGetFramework>? packageTfms) = await GetPackageToUpdateAsync(args.Packages, dgSpec.Projects.Single(), sourceProvider, settings, sourceCacheContext, logger, cancellationToken);
+
+            if (packageToUpdate is null)
+            {
+                // GetPackagesToUpdateAsync is responsible for outputting an error message.
+                return ExitCodes.Error;
+            }
+
+            // 3. Preview restore to validate changes
+            var updatedPackageSpec = dgSpec.Projects[0].Clone();
+            PackageSpecOperations.AddOrUpdateDependency(updatedPackageSpec, packageToUpdate);
+
+            var updatedDgSpec = dgSpec.WithReplacedSpec(updatedPackageSpec).WithoutRestores();
+            updatedDgSpec.AddRestore(updatedPackageSpec.RestoreMetadata.ProjectUniqueName);
+
+            logger.LogDebug("Running Restore preview");
+
+            var restorePreviewResult = await PreviewUpdatePackageReferenceAsync(updatedDgSpec, sourceCacheContext, logger, cancellationToken);
+
+            logger.LogDebug("Restore Review completed");
+
+            if (!restorePreviewResult.Result.Success)
+            {
+                logger.LogMinimal("Restore with updated packages was not successful. Force mode not yet implemented", ConsoleColor.Red);
+                return ExitCodes.Error;
+            }
+
+            var libraryDependency = AddPackageReferenceCommandRunner.GenerateLibraryDependency(
+                updatedPackageSpec,
+                customPackagesPath: null,
+                restorePreviewResult,
+                packageTfms,
+                packageToUpdate);
+
+            // 4. Update MSBuild files
+            if (packageTfms!.Count == dgSpec.Projects[0].TargetFrameworks.Count)
+            {
+                // package is used by all project TFMs (no condition)
+
+                msbuild.AddPackageReference(dgSpec.Projects[0].FilePath, libraryDependency, noVersion: true);
```

Comment:
Does this MSBuild API of ours support CPM?

Reply:
It looks like it tries to. But today we had a customer file a bug saying that `dotnet add package` doesn't support CPM, so I'm not 100% sure.

An earlier version of my changes, before I commit, I explicitly tested for CPM and had an error message saying not yet supported. But when I saw the `msbuild.AddPackageReference` code that apepars to support CPM, I removed it.

I'll try to find time today to do a basic test, but in any case, there's also future improvements planned:
* https://github.com/NuGet/Home/issues/14318